### ISSUE TITLE
Improve Model

### ISSUE THREAD

**Issue by sam-writer** [CodeForPhilly/chime#99]
## Overview

Currently, we use a deterministic SIR model (see `sir` and `sim_sir` in [models.py](https://github.com/CodeForPhilly/chime/blob/develop/penn_chime/models.py)) to predict everything. It does not have many parameters, which I think contributes to the ease of use and adoption of the tool... however, accuracy is also of paramount importance. There have been multiple proposed improvements.

## Proposed Improvements

* [This repo](https://github.com/twiecki/covid19) uses MCMC sampling to do more probabilistic models. @sam-qordoba is trying to get the Bayesian SIR model working, but it has [obscure requirements](https://github.com/twiecki/covid19/issues/14). Here is a [colab notebook of the main model](https://colab.research.google.com/drive/1yALbtOk4x7_koKyNt4jW5EgleUBdtuPD) - the repo works but there are a few setup steps
* Paper suggestion via Google AI: [Bayesian Models for Heterogeneous Personalized Health Data](https://arxiv.org/abs/1509.00110) - [src](https://codeforphilly.slack.com/archives/C01080RBV3Q/p1584635931101200)
* Possibly-useful Transformer model from Google: [interpretable multi-horizon forecasting with deep learning](https://github.com/google-research/google-research/tree/master/tft), but, " unfortunately for this task at the moment, the amount of data seems very limited and expert human biases (e.g. it takes X days to show symptoms/recover etc.) seem more important." [src](https://codeforphilly.slack.com/archives/C01080RBV3Q/p1584567838085000)
* Model should incorporate potential incoming infections from neighboring areas (etc), rather than the assumption of jurisdiction lockdown

## Concerns

* "my understanding is the SIR model's more of a guesstimate that can be fit retrospectively but isn't that predictive for changing circumstances. It doesn't account for household contact, or hordes of folks driving their dying relatives from one jam-packed hospital to the next, or the larger consequences of jamming 200 octogenarians into a group home manned by underpaid attendants with a shortage of tests and protective gear. But we're fighting the epidemic blind, so it's what we've got." [src](https://codeforphilly.slack.com/archives/C01080RBV3Q/p1584590737094600?thread_ts=1584568339.087600&cid=C01080RBV3Q)
* "There has been a lot of talk about using more complex models, but the hurdles are (1) usability (2) uncertainty/unavailability of the required inputs.  I think that the consensus is that better models would be better if they had well-constrianed inputs and didn't make the tool harder for users to adapt to their local contexts.  Otherwise better models would be worse." [src](https://codeforphilly.slack.com/archives/C01080RBV3Q/p1584637089101400?thread_ts=1584635931.101200&cid=C01080RBV3Q)
* "What are some additional inputs that are missing? Am very interested in modelling with different values for various activities, e.g. school transmission, intrahousehold transmission, workplace transmission. Obviously this is not straightforward. Moreover the SIR model totally leaves out the fact that different populations are more or less vulnerable, if one is looking at hospital capacity there are pretty sizeable regional population variations. For a first order guess it's useful, but the real world is full of cascading impacts that are very hard to guess." [src](https://codeforphilly.slack.com/archives/C01080RBV3Q/p1584638113104200?thread_ts=1584635931.101200&cid=C01080RBV3Q)

## Definition of Done

This ticket is complete when we have a plan for improving the model, which takes into account the concerns.

**Comment by briglass** [comment]
Regarding incoming infections, this simulator: http://covidsim.eu/ has an "Infections from outside of the population [per day]", which I am assuming to be accomplished via a monotonic daily increase to S, although I am not sure

**Comment by soarik** [comment]
Thee seems to be two major directions:

i) Making SIR modeling more accurate: Approaches may include:
- Integrating Bayesian modeling, as in https://github.com/twiecki/covid19/blob/master/covid19_growth_bayes.ipynb
- Considering parametric modeling along with black-box ML to fit some portion of such models, as done in physics etc.: https://arxiv.org/pdf/2001.04385.pdf,
- Integrating interventions into modeling
- Proposing some other ways of integrating human expert inductive biases into ML model design (e.g. heavy regularization or designing architectures that use information only via particular paths) so that the past data can be used in better way while avoiding overfitting.

ii) Integrating extra data:
- Maps data for social distancing and mobility estimation, as in this study: https://www.economist.com/graphic-detail/2020/03/14/foot-traffic-has-fallen-sharply-in-cities-with-big-coronavirus-outbreaks
- Search logs for signs of disease (e.g. "medicine for fever" etc.)
- Social network data to reflect potential spreading paths

For (i), there are challenges around model building and validation. For (ii), there are challenges around feasibility of data collection given privacy/security regulations etc.

**Comment by hakocon** [comment]
I really like the idea of social distancing (or I was also thinking population density as a proxy based on ZIP or something like that). Also, would it be valuable to bring in the age distribution of a location? This is specifically dangerous for older folks, so locations with a higher number of elderly might see different distributions of acute/critical care patients than those with lower numbers. Just my thoughts, thanks!

**Comment by sam-writer** [comment]
@hakocon demographics seem very important, I agree, and that is probably just an API call away.

Population distancing as a proxy for social distancing, however, I feel like that assumption could be very wrong - e.g. San Fransisco, with it's shelter in place order, vs somewhere in rural Alaska which, as of today, [has a Representative who is saying](https://www.adn.com/politics/2020/03/19/rep-don-young-tells-alaska-seniors-beer-virus-fears-are-overblown-skips-vote-on-covid-19-relief/) that fears over "the beer virus" are "blown out of proportion" and that "it’s not nearly as deadly as the other viruses we have."

I don't think we should infer this... but if we were going to automatically infer social distancing for a location, I would rather do it by an algorithm like (pseudocode):

```python
social_distancing_phrases = ["shelter in place", "social distancing", ...]
hoax_phrases = ["hoax", "overblown", ...]

sd_score = 0
hoax_score = 0

def infer_social_distancing(city_name):
    search_results = get_google_top_100_pages_for_term(city_name + " coronavirus")
    texts = get_text_from_pages(search_results)
    for sdp in social_distancing_phrases:
        if sdp in text:
            sd_score += 1
        if hoax_score in text:
            hoax_score += 1
    score = somehow_convert_these_to_a_score(sd_score, hoax_score)
    return score
```

**Comment by selahlynch** [comment]
Are daily counts of COVID related hospital/ICU bed usage at Penn Medicine or another hospital system available?  If so we can use this to test some of the models.

**Comment by cjbayesian** [comment]
We need to have an improved way to estimate current # infected. I like the approach taken here: https://github.com/jwrichar/COVID19-mortality/blob/master/State-wise%20Unreported%20COVID-19%20Cases%20in%20the%20U.S..ipynb

**Comment by quinn-dougherty** [comment]
issue too broad, but has a lot of useful notes and discussion on it so i'm reticent to close it

**Comment by sam-writer** [comment]
@quinn-dougherty I agree, this is too broad, but the collection of ideas is valuable. Is there some way to make the output of this ticket:
* an artifact, summarizing these ideas and the ideas on slack? Not sure what that would look like, or if this issue is possibly the best format for an ongoing brainstorming session on model improvement...
* a ticket, explaining what model improvements are to be done at this time?
